### ISSUE TITLE
Type argument inference doesn't seem to work correctly with generic type aliases

### ISSUE THREAD

**Issue by seanchas116** [Microsoft/TypeScript#5417]
Hi, I found that the following example that uses type argument inference and generic type alias doesn't compile:

### Example with generic type alias (does not work)

```ts
type Maybe<T> = T | void;

function get<T>(x: Maybe<T>): T {
	return null; // just an example
}

let foo: Maybe<string>;
get(foo).toUpperCase(); // I think T == string in this function call...
```

#### Error

```
test.ts(8,10): error TS2339: Property 'toUpperCase' does not exist on type 'string | void'.
```

### Example with generic interface (works)

```ts
interface Maybe<T> {}

function get<T>(x: Maybe<T>): T {
	return null;
}

let foo: Maybe<string>;
get(foo).toUpperCase();
```

### Version

```
1.8.0-dev.20151027
```

**Comment by mhegazy** [comment]
The problem is not type alaises. the issue here is that `foo` is of type `string|void`. here is a simpler example:

```ts
function get<T>(x: T | void): T {
    return null; // just an example
}

let foo: string | void;
get(foo).toUpperCase(); // T is still string|void
```

**Comment by seanchas116** [comment]
Thanks! I now see the meaning of the error message.
But isn't this a inconsistent behavior between generic type aliases and generic interfaces/classes?

**Comment by seanchas116** [comment]
Actually, I tried this example first to achieve simple non-nullability:

```ts
type Maybe<T> = T | void;

function exists<T>(x: Maybe<T>): x is T {
    return x != null;
}

let foo: Maybe<string>;
foo.toUpperCase(); // error
if (exists(foo)) {
  foo.toUpperCase(); // OK
}
```

**Comment by mhegazy** [comment]
clever :) a word of caution, trying to wrestle in non-nullable types in the system might be a tedious and brittle thing to do. For future references implementing non-nullable types is tracked by https://github.com/Microsoft/TypeScript/issues/185.

**Comment by ahejlsberg** [comment]
Now fixed by #5738.
